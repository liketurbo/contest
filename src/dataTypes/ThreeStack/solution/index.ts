const NUMBER_OF_STACKS = 3

class ThreeStack<T> {
  private storage: T[]
  private sizes: number[]

  constructor() {
    this.storage = []
    this.sizes = new Array(NUMBER_OF_STACKS).fill(0)
  }

  private getCurIdx(idx: ThreeStackIdx) {
    return Math.max(this.getSize(idx) - 1, 0) * NUMBER_OF_STACKS + idx
  }

  peek(idx: ThreeStackIdx) {
    const curIdx = this.getCurIdx(idx)

    return this.storage[curIdx]
  }

  pop(idx: ThreeStackIdx) {
    const curIdx = this.getCurIdx(idx)

    this.sizes[idx]--

    const value = this.storage[curIdx]

    delete this.storage[curIdx]

    return value
  }

  push(idx: ThreeStackIdx, value: T) {
    const nextIdx = this.getSize(idx) * NUMBER_OF_STACKS + idx

    this.storage[nextIdx] = value

    this.sizes[idx]++

    return this.sizes[idx]
  }

  getSize(idx: ThreeStackIdx) {
    return this.sizes[idx]
  }
}

export type ThreeStackIdx = 0 | 1 | 2

export default ThreeStack
