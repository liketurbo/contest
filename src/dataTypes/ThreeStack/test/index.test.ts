import ThreeStack from "../solution"

describe(ThreeStack, () => {
  const threeStack = new ThreeStack()

  threeStack.push(0, "a")
  threeStack.push(0, "b")
  threeStack.push(0, "c")

  threeStack.push(1, "d")
  threeStack.push(1, "e")

  threeStack.push(2, "f")

  test("push works", () => {
    expect(threeStack).toMatchSnapshot()
  })

  test("peek works", () => {
    expect(threeStack.peek(0)).toBe("c")
    expect(threeStack.peek(1)).toBe("e")
    expect(threeStack.peek(2)).toBe("f")
  })

  test("pop works", () => {
    expect(threeStack.pop(0)).toBe("c")
    expect(threeStack.pop(1)).toBe("e")
    expect(threeStack.pop(2)).toBe("f")
  })
})
