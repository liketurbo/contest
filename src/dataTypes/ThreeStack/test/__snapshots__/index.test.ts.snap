// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeStack push works 1`] = `
ThreeStack {
  "sizes": Array [
    3,
    2,
    1,
  ],
  "storage": Array [
    "a",
    "d",
    "f",
    "b",
    "e",
    undefined,
    "c",
  ],
}
`;
